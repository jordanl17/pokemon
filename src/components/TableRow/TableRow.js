import React from "react";
import PropTypes from "prop-types";

import MUITableRow from "@material-ui/core/TableRow";
import TableCell from "@material-ui/core/TableCell";
import Avatar from "@material-ui/core/Avatar";

import EvolutionsPopup from "../EvolutionsPopup";

import determineRelevantColumns from "../../helpers/columnFilter";
import { alignment } from "../../helpers/alignment";
import pokemonPropTypes from "../../pokemonPropTypes";

const NEXT_EVOLUTION = "next_evolution";

const renderCell = (pokemon, cellName) => {
  /**
   * if currently rendering the next evolution cell, and this pokemon has a next evolution value
   * - show the button
   */
  if (cellName === NEXT_EVOLUTION && pokemon[NEXT_EVOLUTION]) {
    return <EvolutionsPopup pokemon={pokemon} />;
  }
  if (Array.isArray(pokemon[cellName])) {
    // if cell data is an array split each el onto a new line
    return pokemon[cellName].map(item => <div key={item}>{item}</div>);
  }

  if (cellName === "img") {
    return <Avatar alt={pokemon.name} src={pokemon.img} />;
  }
  return pokemon[cellName];
};

const TableRow = ({ isPopup, pokemon }) => (
  <MUITableRow key={pokemon.num}>
    {determineRelevantColumns(isPopup).map(({ name: cellName }, index) => (
      <TableCell
        data-testid={`pokemon-${cellName}`}
        component="th"
        scope="row"
        align={alignment(index)}
        key={cellName}
      >
        {renderCell(pokemon, cellName)}
      </TableCell>
    ))}
  </MUITableRow>
);

TableRow.propTypes = {
  isPopup: PropTypes.bool.isRequired,
  pokemon: pokemonPropTypes
};

export default TableRow;
