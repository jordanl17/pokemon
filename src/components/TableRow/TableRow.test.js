import React from "react";
import { render } from "@testing-library/react";

import TableRow from "./TableRow";

const defaultProps = {
  isPopup: false,
  pokemon: {
    name: "test name",
    num: "test num",
    type: "test type",
    weight: "test weight",
    height: "test height"
  }
};
const renderer = props =>
  render(
    <table>
      <tbody>
        <TableRow {...defaultProps} {...props} />
      </tbody>
    </table>
  );
describe("TableRow", () => {
  it("should render the correct rows when displaying a non popup table", () => {
    const { getByText } = renderer({ isPopup: true });

    getByText("test name");
    getByText("test num");
    getByText("test type");
    getByText("test weight");
    getByText("test height");
  });

  it("shouldn't render a next evolution button when there are no next evolutions on this pokemon", () => {
    const { queryByTestId } = renderer();

    expect(queryByTestId("evolution-button")).toBeNull();
  });

  it("shouldn't render a next evolution button when pokemon has next evolution but table is as popup", () => {
    const { queryByTestId } = renderer({
      isPopup: true,
      pokemon: {
        next_evolutions: "test next"
      }
    });

    expect(queryByTestId("evolution-button")).toBeNull();
  });

  it("should render next evolutions button when there is a next evolution on this pokemon", () => {
    const { queryByTestId } = renderer({
      pokemon: { next_evolutions: "test next" }
    });

    queryByTestId("evolution-button");
  });

  it("should render an array cell as a series of divs", () => {
    const { getByText } = renderer({
      pokemon: { type: ["first type", "second type"] }
    });

    const typeCell = getByText("first type").parentElement;
    expect(typeCell).toMatchSnapshot();
  });
});
