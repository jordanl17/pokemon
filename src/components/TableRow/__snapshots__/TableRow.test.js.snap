// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should render an array cell as a series of divs 1`] = `
<th
  class="MuiTableCell-root MuiTableCell-alignRight"
  data-testid="pokemon-type"
  scope="row"
>
  <div>
    first type
  </div>
  <div>
    second type
  </div>
</th>
`;
